networkx/algorithms/swap.py

```diff
@@ -6,7 +6,130 @@
+       A directed graph
+
+    nswap : integer (optional, default=1)
+       Number of three-edge swaps to perform
+
+    max_tries : integer (optional)
+       Maximum number of attempts to swap edges
+
+    seed : integer, random_state, or None (default)
+        Indicator of random number generation state.
+        See :ref:`Randomness<randomness>`.
+
+    Returns
+    -------
+    G : DiGraph
+       The graph after triple-edge swaps.
+
+    Raises
+    ------
+    NetworkXError
+        If `G` is not directed, or
+        If nswap > max_tries, or
+        If there are fewer than 4 nodes in `G`
+    NetworkXAlgorithmError
+        If the number of swap attempts exceeds `max_tries` before `nswap` swaps are made
+
+    Notes
+    -----
+    Does not enforce any connectivity constraints.
+
+    The graph G is modified in place.
+
+    References
+    ----------
+    .. [1] Erdős, Péter L., et al. “A Simple Havel-Hakimi Type Algorithm to Realize
+           Graphical Degree Sequences of Directed Graphs.” ArXiv:0905.4913 [Math],
+           Jan. 2010. https://doi.org/10.48550/arXiv.0905.4913.
+    .. [2] “Combinatorics - Reaching All Possible Simple Directed Graphs with a given
+           Degree Sequence with 2-Edge Swaps.” Mathematics Stack Exchange,
+           https://math.stackexchange.com/questions/22272/. Accessed 30 May 2022.
+    """
+    if not G.is_directed():
+        raise nx.NetworkXError(
+            "directed_edge_swap() is only defined for directed graphs. Use double_edge_swap "
+            "instead"
+        )
+    if nswap > max_tries:
+        raise nx.NetworkXError("Number of swaps > number of tries allowed.")
+    if len(G) < 4:
+        raise nx.NetworkXError("Graph has less than four nodes.")
+
+    # Instead of choosing uniformly at random from a generated edge list,
+    # this algorithm chooses nonuniformly from the set of nodes with
+    # probability weighted by degree.
+    n = 0
+    swapcount = 0
+    keys, degrees = zip(*G.degree())  # keys, degree
+    cdf = nx.utils.cumulative_distribution(degrees)  # cdf of degree
+    discrete_sequence = nx.utils.discrete_sequence
+
+    while swapcount < nswap:
+        # choose source node index from discrete distribution
+        start_index = discrete_sequence(1, cdistribution=cdf, seed=seed)[0]
+        start = keys[start_index]
+
+        # If the given node doesn't have any out edges, then there isn't anything to swap
+        if G.out_degree(start) == 0:
+            n += 1
+            continue
+        _, second = seed.choice(list(G.out_edges(start)))
+
+        if G.out_degree(second) == 0:
+            n += 1
+            continue
+        _, third = seed.choice(list(G.out_edges(second)))
+
+        if G.out_degree(third) == 0:
+            n += 1
+            continue
+        _, fourth = seed.choice(list(G.out_edges(third)))
+
+        if start == second or second == third or third == fourth:
+            n += 1
+            continue
+
+        if (
```

Comment:
Similarly here, use `G.succ` rather than `out_edges to check for successors.  This time you don’t want a list of the successors, a dict is better. So use `G.succ` Rather than this dict-of-lists `nbrs
```python
If third not in G.succ[start] and fourth not in G.succ[second] and second not in G.succ[third]:
```
You can check the timing of whether it is significant to create a variable to hold `G.succ`.  Sometimes it is better to define at the top of the function `G_succ = G.succ` to avoid the same lookup over and over.

Reply:
Looks like `G_succ = G.succ` didn't make much of a difference here. The original took ~262 seconds, and the  `G_succ = G.succ` version took ~267. 
``` python
import copy
import networkx as nx
import time

graph = nx.gnp_random_graph(2000, 0.5, directed=True)
graph2 = copy.deepcopy(graph)

start = time.time()
G = nx.directed_edge_swap(graph, nswap=1000000, max_tries=1e16, seed=1)
end = time.time()
print(end - start)
start = time.time()
G = nx.other_edge_swap(graph2, nswap=1000000, max_tries=1e16, seed=1)
end = time.time()
print(end - start)

```